zipkin/src/main/java/com/linecorp/armeria/common/tracing/RequestContextCurrentTraceContext.java

```diff
@@ -0,0 +1,127 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.tracing;
+
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.internal.tracing.PingPongExtra;
+
+import brave.propagation.CurrentTraceContext;
+import brave.propagation.TraceContext;
+import io.netty.util.Attribute;
+import io.netty.util.AttributeKey;
+
+/**
+ * {@linkplain brave.Tracing.Builder#currentTraceContext(brave.propagation.CurrentTraceContext) Tracing
+ * context} implemented with {@linkplain com.linecorp.armeria.common.RequestContext}.
+ *
+ * <p>This {@link CurrentTraceContext} stores/loads the trace context into/from a
+ * {@link RequestContext}'s attribute so that there's no need for thread local variables
+ * which can lead to unpredictable behavior in asynchronous programming.
+ */
+public final class RequestContextCurrentTraceContext extends CurrentTraceContext {
+
+    public static final CurrentTraceContext INSTANCE = new RequestContextCurrentTraceContext();
+    private static final Logger logger = LoggerFactory.getLogger(RequestContextCurrentTraceContext.class);
+    private static final AttributeKey<TraceContext> TRACE_CONTEXT_KEY =
+            AttributeKey.newInstance("trace-context");
+
+    /**
+     * Use this to ensure the trace context propagates to children.
+     *
+     * <p>Ex.
+     * <pre>{@code
+     *  // Ensure the trace context propagates to children
+     * ctx.onChild(RequestContextCurrentTraceContext::copy);
+     * }</pre>
+     */
+    public static void copy(RequestContext src, RequestContext dst) {
+        dst.attr(TRACE_CONTEXT_KEY).set(src.attr(TRACE_CONTEXT_KEY).get());
+    }
+
+    @Override
+    public TraceContext get() {
+        if (hasCurrentRequestContextOrWarnOnce()) {
+            return null;
+        }
+        return RequestContext.current().attr(TRACE_CONTEXT_KEY).get();
+    }
+
+    @Override
+    public CurrentTraceContext.Scope newScope(TraceContext currentSpan) {
+        // Handle inspection added to ensure we can fail-fast if this isn't installed.
+        if (PingPongExtra.maybeSetPong(currentSpan)) { return Scope.NOOP; }
+
+        if (hasCurrentRequestContextOrWarnOnce()) {
+            return Scope.NOOP;
+        }
+
+        final Attribute<TraceContext> contextAttribute = RequestContext.current().attr(TRACE_CONTEXT_KEY);
+        final TraceContext previous = contextAttribute.get();
+        contextAttribute.set(currentSpan);
+
+        // Don't remove the outer-most context (client or server request)
+        if (previous == null) { return Scope.NOOP; }
+
+        // Allow sub-spans (aka local spans) to close when Brave's scope does.
```

Comment:
`// Removes sub-spans (i.e. local spans) from the current context when Brave's scope does. If an asynchronous sub-span, it may still complete later.`